src/threshold_sign.rs

```diff
@@ -100,26 +104,39 @@ impl<N: NodeIdT> DistAlgorithm for ThresholdSign<N> {
 
 impl<N: NodeIdT> ThresholdSign<N> {
     /// Creates a new instance of `ThresholdSign`, with the goal to collaboratively sign `msg`.
-    pub fn new<M: AsRef<[u8]>>(netinfo: Arc<NetworkInfo<N>>, msg: M) -> Self {
+    pub fn new(netinfo: Arc<NetworkInfo<N>>) -> Self {
         ThresholdSign {
             netinfo,
-            msg_hash: hash_g2(msg),
+            doc_hash: None,
             received_shares: BTreeMap::new(),
             had_input: false,
             terminated: false,
         }
     }
 
+    /// Sets doc_hash. Signature shares can only be sent after this function is completed.
+    pub fn set_document<M: AsRef<[u8]>>(&mut self, msg: M) -> Result<()> {
+        if self.doc_hash.is_some() {
+            return Err(Error::MultipleMessagesToSign);
+        }
+        self.doc_hash = Some(hash_g2(msg));
+        Ok(())
+    }
+
     /// Sends our signature shares, and if we have collected enough, returns the full signature.
+    /// Returns an empty step if the message to sign hasn't been received yet.
     pub fn sign(&mut self) -> Result<Step<N>> {
-        if self.had_input {
-            return Ok(Step::default());
-        }
+        let hash = match self.doc_hash {
+            Some(hash) => hash,
+            None => return Err(Error::DocumentHashIsNone),
+        };
         self.had_input = true;
+        let mut step = Step::default();
+        step.fault_log.extend(self.remove_invalid_shares());
```

Comment:
We need to make sure this step isn't dropped. It needs to say

```rust
return step.join(self.try_output());
```

in line 137 and

```rust
step.messages.push(Target::All.message(msg.clone()));
```

in line 140 now.

Reply:
ah ouch, I presume if the tests weren't failing when I committed they would have caught this? If not, maybe I should write a test to try and catch something like this?

Reply:
I know, our tests still leave a lot to be desired. (We currently also have two "network test frameworks", and need to port everything to the new one.)
They don't test that faults are actually reported at all yet. See #120—it would be great to make a start on that one! :+1: